internal/events/event.go

```diff
@@ -355,32 +355,30 @@ func (e Events) prepareEventsTable(ctx context.Context, tx *gorm.DB, clusterID *
 	//host bound to a cluster or registered to a bound infra-env) check the access permission
 	//relative to the cluster ownership
 	if clusterBoundEvents() {
-		tx = tx.Model(&common.Cluster{}).
-			Select("events.*, clusters.user_name, clusters.org_id").
-			Joins("INNER JOIN events ON clusters.id = events.cluster_id")
+		tx = tx.Model(&common.Event{}).Select("events.*, clusters.user_name, clusters.org_id").
+			Joins("INNER JOIN clusters ON clusters.id = events.cluster_id")
 
 		// if deleted hosts flag is true, we need to add 'deleted_at' to know whether events are related to a deleted host
 		if swag.BoolValue(deletedHosts) {
 			tx = tx.Select("events.*, clusters.user_name, clusters.org_id, hosts.deleted_at").
-				Joins("LEFT JOIN hosts ON events.host_id = hosts.id")
+				Joins("LEFT JOIN hosts ON hosts.id = events.host_id")
 		}
 		return tx
 	}
 
 	//for unbound events that are searched with infra-env id (whether events on hosts or the
 	//infra-env level itself) check the access permission relative to the infra-env ownership
 	if nonBoundEvents() {
-		return tx.Model(&common.InfraEnv{}).
-			Select("events.*, infra_envs.user_name, infra_envs.org_id").
-			Joins("INNER JOIN events ON infra_envs.id = events.infra_env_id")
+		return tx.Model(&common.Event{}).Select("events.*, infra_envs.user_name, infra_envs.org_id").
+			Joins("INNER JOIN infra_envs ON infra_envs.id = events.infra_env_id")
 	}
 
-	//for query made on the host only check the permission relative to it's infra-env. since
-	//host table does not contain an org_id we can not perform a join on that table and has to go
-	//through the infra-env table which is good because authorization is done on the infra-level
+	// Events must be linked to the infra_envs table and then to the hosts table
+	// The hosts table does not hold an org_id, so permissions related fields must be supplied by the infra_env
 	if hostOnlyEvents() {
-		return tx.Model(&common.Host{}).Select("events.*, infra_envs.user_name, infra_envs.org_id").
-			Joins("INNER JOIN infra_envs ON hosts.infra_env_id = infra_envs.id").Joins("INNER JOIN events ON events.host_id = hosts.id")
+		return tx.Model(&common.Event{}).Select("events.*, infra_envs.user_name, infra_envs.org_id").
+			Joins("INNER JOIN infra_envs ON infra_envs.id = events.infra_env_id").
+			Joins("INNER JOIN hosts ON hosts.id = events.host_id")
```

Comment:
why are we joining on the host table? As we are not selecting, the only effect we would have is not returning any event where `events.host_id` has no match in the `hosts` table. Am I missing something here?

Reply:
The event must have a match in the hosts table + a match in the infra envs table. This is the purpose of this join. We don't need fields from the host for this as you can see in the `Select` nothing has changed.

I think this is reasonable, the other option is to do a `WHERE events.host_id IN ...` type of subquery but I don't think this provides any benefit. Arguably we return less data I guess?

WDYT do you think we should change this to check for an existing host by using a subquery?

Reply:
I think it's good as it is, maybe let's make the comment slightly more explicit ("ensuring host is there" or something along those lines, rather than "linking to host table")

Reply:
Good point.

I added a comment `// This join is here to ensure that only events for a host that exists are fetched` to make the intent of the join without select a bit clearer.